PsfLauncher/main.cpp

```diff
@@ -265,46 +277,259 @@ void LaunchMonitorInBackground(std::filesystem::path packageRoot, const wchar_t
-            0, // CreationFlags
-            nullptr, // Environment
+        return error;
+    }
+
+    return {};
+}
+
+ErrorInformation StartProcess(ExecutionInformation execInfo, int cmdShow, bool runInVirtualEnvironment) noexcept
+{
+    STARTUPINFOEXW startupInfoEx =
+    {
+        {
+        sizeof(startupInfoEx)
+        , nullptr //lpReserved
+        , nullptr // lpDesktop
+        , nullptr // lpTitle
+        , 0 //dwX
+        , 0 //dwY
+        , 0 //dwXSize
+        , 0 //swYSize
+        , 0 //dwXCountChar
+        , 0 //dwYCountChar
+        , 0 // dwFillAttribute
+        , STARTF_USESHOWWINDOW //dwFlags
+        , static_cast<WORD>(cmdShow) // wShowWindow
+        }
+    };
+
+    if (runInVirtualEnvironment)
+    {
+        SIZE_T AttributeListSize;
+        InitializeProcThreadAttributeList(nullptr, 1, 0, &AttributeListSize);
+        startupInfoEx.lpAttributeList = (LPPROC_THREAD_ATTRIBUTE_LIST)HeapAlloc(
+            GetProcessHeap(),
+            0,
+            AttributeListSize
+        );
+
+        if (InitializeProcThreadAttributeList(startupInfoEx.lpAttributeList,
+            1,
+            0,
+            &AttributeListSize) == FALSE)
+        {
+            auto err{ ::GetLastError() };
+            ErrorInformation error = { L"Could not initialize the proc thread attribute list.", err };
+        }
+
+        DWORD attribute = PROCESS_CREATION_DESKTOP_APP_BREAKAWAY_DISABLE_PROCESS_TREE;
+        if (UpdateProcThreadAttribute(startupInfoEx.lpAttributeList,
+            0,
+            PROC_THREAD_ATTRIBUTE_DESKTOP_APP_POLICY,
+            &attribute,
+            sizeof(attribute),
             nullptr,
-            &startupInfo,
-            &processInfo))
+            nullptr) == FALSE)
         {
-            if (::GetLastError() == ERROR_ELEVATION_REQUIRED)
-                Log("error starting monitor using CreateProcessW. You must specify 'monitor/asadmin' in config.json\n");
-            else
-                Log("error starting monitor using CreateProcessW. Error=0x%x\n", ::GetLastError());
+            auto err{ ::GetLastError() };
+            ErrorInformation error{ L"Could not update Proc thread attribute.", err };
+        }
+    }
+
+    PROCESS_INFORMATION processInfo = { 0 };
+    if (::CreateProcessW(
+        execInfo.ApplicationName,
+        execInfo.CommandLine.data(),
+        nullptr, nullptr, // Process/ThreadAttributes
+        true, // InheritHandles
+        EXTENDED_STARTUPINFO_PRESENT, // CreationFlags
+        nullptr, // Environment
+        execInfo.CurrentDirectory,
+        (LPSTARTUPINFO)&startupInfoEx,
+        &processInfo))
+    {
+        // Propagate exit code to caller, in case they care
+        DWORD waitResult = ::WaitForSingleObject(processInfo.hProcess, INFINITE);
+
+        if (waitResult != WAIT_OBJECT_0)
+        {
+            auto err{ ::GetLastError() };
+            return { L"Waiting operation failed unexpectedly.", err };
+        }

```

Comment:
not propagating exit code?

Reply:
That was an old comment from Tim.  That is removed.